core/indigo-core/molecule/src/molecule_json_saver.cpp

```diff
@@ -1370,34 +1523,225 @@ void MoleculeJsonSaver::saveMolecule(BaseMolecule& bmol, JsonWriter& writer)
         writer.EndObject();
     }
 
+    int mon_idx = 0;
+
+    // save references to monomer's instances
+    for (auto i : mol.vertices())
+    {
+        if (mol.isTemplateAtom(i))
+        {
+            writer.StartObject();
+            writer.Key("$ref");
+            writer.String((std::string("monomer") + std::to_string(mon_idx)).c_str());
+            writer.EndObject();
+            _monomers_enum.emplace(i, mon_idx++);
+        }
+    }
+
+    for (const auto& kvp : _scsr_superatoms)
+    {
+        writer.StartObject();
+        writer.Key("$ref");
+        writer.String((std::string("monomer") + std::to_string(mon_idx + kvp.second)).c_str());
+        writer.EndObject();
+    }
+
     writer.EndArray(); // nodes
 
-    if (mol->tgroups.getTGroupCount())
+    if (mol.tgroups.getTGroupCount())
     {
+        // collect attachment points into unordered map <key, val>. keu - pair of from and destination atom. val - attachment point name.
+        _monomer_connections.clear();
+        for (int i = mol.template_attachment_points.begin(); i != mol.template_attachment_points.end(); i = mol.template_attachment_points.next(i))
+        {
+            auto& sap = mol.template_attachment_points.at(i);
+            _monomer_connections.emplace(std::make_pair(sap.ap_occur_idx, sap.ap_aidx), sap.ap_id.ptr());
+        }
+
+        for (const auto& kvp : _scsr_superatoms)
```

Comment:
there should be no _scsr_superatoms
all the modified monomers should be loaded into TGroups while loading from molfile